### PR TITLE
Add dataset api support

### PR DESCRIPTION
**Pull request by rclark**
Adds basic dataset api support:

```
mapbox dataset list
mapbox dataset create
mapbox dataset read-dataset
mapbox dataset delete-dataset
mapbox dataset delete-dataset
mapbox dataset list-features
mapbox dataset read-feature
mapbox dataset put-feature
mapbox dataset delete-feature
mapbox dataset batch-update-feature
mapbox dataset create-tileset
```

The only deviation from the sdk was naming `put-feature` rather than `update-feature`. It seems a little confusing to me to have a method called `update-feature` where the first line of documentation says, "Insert or update". 

Also implements `mapbox dataset create-tileset` which is an `uploads.create` call in disguise.

cc @perrygeo @sgillies for any further tips on writing python code!

refs #21

### CODE REVIEW

**Comment by rclark** [comment]
per https://github.com/mapbox/mapbox-cli-py/issues/21#issuecomment-165244203, I would also like to add three "higher-level" functions to `mapbox datasets`. Objections?

File: mapboxcli/scripts/datasets.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+@dataset.command(name="batch-update-feature",
+    short_help="Insert, update, or delete multiple features in a dataset")
+@click.argument('dataset', required=True)
+@click.argument('puts', required=False, default=None)
+@click.argument('deletes', required=False, default=None)
+@click.option('--input', '-i', default='-',
+    help="File containing features to insert, update, and/or delete")
+@click.pass_context
+def delete_feature(ctx, dataset, puts, deletes, input):
```
Comment (perrygeo) [inline mapboxcli/scripts/datasets.py]: s/delete_feature/batch_update_feature/

**Comment by perrygeo** [comment]
@rclark This looks greats.

A few questions

* The interface to `batch-update-features` was a bit confusing to me, particularly with optional positional args for puts and deletes. What do you think about breaking that up into `batch-put-features` and `batch-delete-features`?

* The order of subcommands that you suggest above it nice and logical. But click insists on listing it alphabetically which makes the ordering a bit odd. I haven't found any solutions for this yet but maybe others have some ideas? 

* Should we name it `datasets` (plural) to be consistent with the developers API page?

The higher-level commands, chunked writing and geojson feature streaming are something we should definitely keep on the radar. For now though, I'm happy with the current implementation as a first cut.

**Comment by rclark** [comment]
I agree that the batch update interface is a little confusing, but I'm inclined to keep it because it is a mirror image of the actual API endpoint. In the [higher-ground branch](https://github.com/mapbox/mapbox-cli-py/blob/673cf3a7abfabf9b23d8cb04c4c80f2b7e65132f/mapboxcli/scripts/datasets.py#L324-L407) I outlined distinct put-features and delete-features methods which would might be more clear and more amenable to streaming from other geojson sources. 

:+1: to rolling those new methods separately in the near future, perhaps as soon as https://github.com/mapbox/cligj/pull/9 lands and we're in a better place to stream geojson through.

I'm missing something here about order of subcommands. Do you mean that you don't like the fact that `mapbox datasets --help` lists the subcommands in a different order than I wrote them in the code?

:+1: I'll rename to datasets. I think I was thrown by the singular `upload` command.

**Comment by perrygeo** [comment]
@rclark 

re: ordering, i was just noting that in your original comment, the command order makes sense:

```
mapbox dataset list
mapbox dataset create
mapbox dataset read-dataset
mapbox dataset delete-dataset
mapbox dataset update-dataset
mapbox dataset list-features
mapbox dataset read-feature
mapbox dataset put-feature
mapbox dataset delete-feature
mapbox dataset batch-update-feature
mapbox dataset create-tileset
```
But `click` alphabetizes them resulting in a not-very-logical order in the generated help page
```
batch-update-feature 
create               
create-tileset       
delete-dataset       
delete-feature       
list                 
list-features        
put-feature          
read-dataset         
read-feature         
update-dataset       
```
It might just be a matter of configuring click to change the ordering but I'm not sure that option exits (haven't found it in the click docs)

Good catch on the `upload` (singular), we should change that as well. Naming things is hard.

**Comment by rclark** [comment]
> on the upload (singular), we should change that as well.

.... actually... maybe we shouldn't. `mapbox upload` is currently actually a higher-level abstraction over a set of API calls. And in this case I think this isn't a singular noun... its actually a verb O_o

```
mapbox upload rclark.my-map ~/tmp/some-data.geojson
```

**Comment by perrygeo** [comment]
>  maybe we shouldn't. mapbox upload is actually a higher-level abstraction over a set of API calls.

Good point, at the sdk it make sense to mirror the API as closely as possible. But maybe we start diverging from the naming at the CLI where it makes sense. I like the idea of making subcommands more verb-like (geocode instead of geocoding) but it doesn't fit every service. I dunno, I suck at naming things, carry on...

**Comment by sgillies** [comment]
Naming is hard, for sure. I'd love to get a more unix-y feel to the commands – `mapbox datasets ls` and `mapbox datasets rm` instead of `mapbox datasets list` and `mapbox datasets delete-dataset` – but it'd be hard to make that work across all the resource-action combos.

**Comment by perrygeo** [comment]
@rclark We should also document the datasets functionality - aka copy and paste the help into the README.

**Comment by rclark** [comment]
Slapped the docs into readme. How about we merge this? I'm excited to get it integrated with #34

**Comment by sgillies** [comment]
@rclark I'm going to give feature access one more test drive and will merge in ~1 hour. The package is otherwise (change log etc) ready for a release.

File: mapboxcli/scripts/datasets.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+    `datasets:write` scope is required, see `mapbox --help`.
+    """
+
+    service = ctx.obj.get('service')
+    res = service.delete_feature(dataset, fid)
+
+    if res.status_code != 204:
+        raise MapboxCLIException(res.text.strip())
+
+@datasets.command(name="batch-update-feature",
```
Comment (sgillies) [inline mapboxcli/scripts/datasets.py]: `name="batch-update-features"` plural.

**Comment by sgillies** [comment]
Here's what I did to test it out

```
$ mapbox datasets create
$fio cat ~/code/Fiona/tests/data/coutwildrnp.shp | fio collect | jq '{put: .features, delete: []}' | mapbox datasets batch-update-feature ciibvsw8g011ynflz5m91tvht
$mapbox datasets list-features ciibvsw8g011ynflz5m91tvht | geojsonio
```

geojson io

**Pull request merged**